data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/router/RouterCopyRecordStrategyTests.java

```diff
@@ -0,0 +1,128 @@
+        assertThat(recordOut, equalTo(firstRecord));
+        recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
+        Collection<Record> recordsOut = getRecordStrategy.getAllRecords(recordsIn);
+        assertThat(recordsOut.size(), equalTo(recordsIn.size()));
+        assertThat(recordsIn.stream().collect(Collectors.toSet()), equalTo(recordsOut.stream().collect(Collectors.toSet())));
+    }
+
+    @Test
+    void test_with_one_data_flow_pipeline_connector() {
+        DataFlowComponent<PipelineConnector> dataFlowComponent = mock(DataFlowComponent.class);
+        
+        Collection<DataFlowComponent<PipelineConnector>> dataFlowComponents
+            = Collections.singletonList(dataFlowComponent);
+        final RouterGetRecordStrategy getRecordStrategy = createObjectUnderTest(dataFlowComponents);
+        Record firstRecord = recordsIn.iterator().next();
+        Record recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
+        recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
+        Collection<Record> recordsOut = getRecordStrategy.getAllRecords(recordsIn);
+        assertThat(recordsOut.size(), equalTo(recordsIn.size()));
+        assertThat(recordsIn.stream().collect(Collectors.toSet()), equalTo(recordsOut.stream().collect(Collectors.toSet())));
+        /* Repeat the test to see if the conditions holds true */
+        recordsOut = getRecordStrategy.getAllRecords(recordsIn);
+        assertThat(recordsOut.size(), equalTo(recordsIn.size()));
+        assertThat(recordsIn.stream().collect(Collectors.toSet()), equalTo(recordsOut.stream().collect(Collectors.toSet())));
+    }
+
+    @Test
+    void test_with_multiple_data_flow_components_and_pipeline_connectors() {
+        Collection<DataFlowComponent<PipelineConnector>> dataFlowComponents = new ArrayList<>();
+        when(pipelineDataFlowComponent.getComponent()).thenReturn(new PipelineConnector());
+        for (int i = 0; i < 3; i++) {
+            dataFlowComponents.add(pipelineDataFlowComponent);
+        }
+
+        final Map<String, Object> eventMap = new HashMap<>();
+        eventMap.put(UUID.randomUUID().toString(), UUID.randomUUID().toString());
+        final JacksonEvent event = JacksonEvent.builder()
+                                       .withData(eventMap)
+                                       .withEventType("event")
+                                       .build();
+        recordsIn = IntStream.range(0, 10)
+                .mapToObj(i -> new Record<Event>(event))
+                .collect(Collectors.toList());
+        final RouterGetRecordStrategy getRecordStrategy = createObjectUnderTest(dataFlowComponents);
+        Record firstRecord = recordsIn.iterator().next();
+        Record recordOut = getRecordStrategy.getRecord(firstRecord);
+        assertThat(recordOut, equalTo(firstRecord));
```

Comment:
I think you should use `sameInstance` to be very explicit on the behavior.

```
assertThat(recordOut, sameInstance(firstRecord));
```

This only works now because `equals`/`hashCode` is not yet implemented.